modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -331,6 +440,79 @@ public void onFailure(Exception e) {
+            @Override
+            public void onResponse(ForceMergeResponse forceMergeResponse) {
+                logger.info("DLM successfully force merged index [{}]", targetIndex);
+                setForceMergeCompletedTimestamp(targetIndex, listener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                String previousError = errorStore.getError(targetIndex);
+                listener.onFailure(e);
+                // To avoid spamming our logs, we only want to log the error once.
+                if (previousError == null || previousError.equals(errorStore.getError(targetIndex)) == false) {
+                    logger.error(
+                        () -> Strings.format(
+                            "DLM encountered an error trying to force merge index [%s]. DLM will attempt to force merge the index on its "
+                                + "next run.",
+                            targetIndex
+                        ),
+                        e
+                    );
+                }
+            }
+        });
+    }
+
+    /*
+     * This method sets the value of the custom index metadata field "force_merge_completed_timestamp" within the field "dlm" to value. The
+     * method returns immediately, but the update happens asynchronously and listener is notified on success or failure.
+     */
+    private void setForceMergeCompletedTimestamp(String targetIndex, ActionListener<Void> listener) {
+        clusterStateUpdateTaskQueue.submitTask(
+            Strings.format("Adding force merge complete marker to cluster state for [%s]", targetIndex),
+            new DataLifecycleClusterStateUpdateTask(listener) {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    logger.trace("Updating cluster state with force merge complete marker for {}", targetIndex);
+                    IndexMetadata indexMetadata = currentState.metadata().index(targetIndex);
+                    Map<String, String> customMetadata = indexMetadata.getCustomData(DLM_CUSTOM_INDEX_METADATA_KEY);
+                    Map<String, String> newCustomMetadata = new HashMap<>();
+                    if (customMetadata != null) {
+                        newCustomMetadata.putAll(customMetadata);
+                    }
+                    newCustomMetadata.put(FORCE_MERGE_COMPLETED_TIMESTAMP_METADATA_KEY, Long.toString(threadPool.absoluteTimeInMillis()));
+                    IndexMetadata updatededIndexMetadata = new IndexMetadata.Builder(indexMetadata).putCustom(
+                        DLM_CUSTOM_INDEX_METADATA_KEY,
+                        newCustomMetadata
+                    ).build();
+                    Metadata metadata = Metadata.builder(currentState.metadata()).put(updatededIndexMetadata, true).build();
+                    return ClusterState.builder(currentState).metadata(metadata).build();
+                }
+            },
```

Comment:
This only has one implementer, so I'm not sure it really needs to be an `abstract` class. What about making this class concrete, and passing in any pieces that are required (like the `targetIndex`)? We could rename it to something like `UpdateForceMergeCompleteTask` to indicate exactly what it does instead of leaving it for generic behavior. What do you think?

That would also mean it could be independently unit tested at some point.

Reply:
Yeah I can change it. It had two implementations until just recently.